web/app/scripts/custom-reports/custom-report-controller.js

```diff
@@ -0,0 +1,88 @@
+(function () {
+    'use strict';
+
+    /* Custom report view.  Receives filter and aggregation parameters via $stateParams,
+     * gets a report from the API and displays it.
+     *
+     * The HTML for the table body and column header is built as a string here in the controller,
+     * because ng-repeat is too slow.
+     */
+
+    /* ngInject */
+    function CustomReportController($state, $stateParams, Records, AggregationsConfig) {
+        var ctl = this;
+        ctl.dateFormat = 'MMM D, YYYY';
+        init();
+
+        function init() {
+            ctl.loading = true;
+            ctl.params = $stateParams;
+
+            getCategoryLabels().then(function () {
+                return Records.report(ctl.params).$promise;
+            }).then(function (report) {
+                console.log('report ready');
+                ctl.report = report;
+                composeTables();
+                ctl.loading = false;
```

Comment:
ctl.loading = false should go in a finally block, instead of in the then. Consider displaying a small error message if getCategoryLabels fails.